File: python/external-session-storage/redis_sample.py
"""Runnable Redis SessionStore reference sample."""

from __future__ import annotations

import argparse
import asyncio
import base64
import json
import os
import time
import uuid
from pathlib import Path
from typing import cast

from qoder_agent_sdk import (
    QoderAgentOptions,
    ResultMessage,
    SDKMirrorErrorMessage,
    SessionKey,
    SessionListSubkeysKey,
    SessionStore,
    SessionStoreEntry,
    SessionStoreListEntry,
    access_token_from_env,
    delete_session,
    delete_session_via_store,
    get_session_info,
    query,
)
from redis.asyncio import Redis

_MAIN_TRANSCRIPT = "main"


def _encoded(value: str) -> str:
    encoded = base64.urlsafe_b64encode(value.encode()).decode()
    return encoded.rstrip("=")


class RedisSessionStore(SessionStore):
    """Store transcript entries in Redis lists with lightweight indexes."""

    def __init__(
        self,
        client: Redis,
        prefix: str = "qoder:samples:session-storage",
    ) -> None:
        self._client = client
        self._prefix = prefix.rstrip(":")

    async def append(
        self,
        key: SessionKey,
        entries: list[SessionStoreEntry],
    ) -> None:
        if not entries:
            return

        transaction = self._client.pipeline(transaction=True)
        transaction.rpush(
            self._entries_key(key),
            *(json.dumps(entry, separators=(",", ":")) for entry in entries),
        )
        subpath = key.get("subpath")
        if subpath is None:
            transaction.zadd(
                self._sessions_key(key["project_key"]),
                {key["session_id"]: int(time.time() * 1000)},
            )
        else:
            transaction.sadd(self._subkeys_key(key), subpath)
        await transaction.execute()

    async def load(self, key: SessionKey) -> list[SessionStoreEntry] | None:
        values = cast(
            list[str],
            await self._client.lrange(self._entries_key(key), 0, -1),
        )
        if not values:
            return None
        return [cast(SessionStoreEntry, json.loads(value)) for value in values]

    async def list_sessions(
        self,
        project_key: str,
    ) -> list[SessionStoreListEntry]:
        values = cast(
            list[tuple[str, float]],
            await self._client.zrange(
                self._sessions_key(project_key),
                0,
                -1,
                desc=True,
                withscores=True,
            ),
        )
        return [
            {"session_id": session_id, "mtime": int(mtime)}
            for session_id, mtime in values
        ]

    async def list_subkeys(self, key: SessionListSubkeysKey) -> list[str]:
        values = cast(
            set[str],
            await self._client.smembers(self._subkeys_key(key)),
        )
        return sorted(values)

    async def delete(self, key: SessionKey) -> None:
        subpath = key.get("subpath")
        if subpath is not None:
            transaction = self._client.pipeline(transaction=True)
            transaction.delete(self._entries_key(key))
            transaction.srem(self._subkeys_key(key), subpath)
            await transaction.execute()
            return

        subpaths = cast(
            set[str],
            await self._client.smembers(self._subkeys_key(key)),
        )
        keys = [self._entries_key(key), self._subkeys_key(key)]
        keys.extend(
            self._entries_key({**key, "subpath": subpath}) for subpath in subpaths
        )
        transaction = self._client.pipeline(transaction=True)
        transaction.delete(*keys)
        transaction.zrem(
            self._sessions_key(key["project_key"]),
            key["session_id"],
        )
        await transaction.execute()

    def _entries_key(self, key: SessionKey) -> str:
        subpath = key.get("subpath")
        transcript = (
            f"child:{_encoded(subpath)}" if subpath is not None else _MAIN_TRANSCRIPT
        )
        return ":".join(
            [
                self._prefix,
                "entries",
                _encoded(key["project_key"]),
                _encoded(key["session_id"]),
                transcript,
            ]
        )

    def _sessions_key(self, project_key: str) -> str:
        return ":".join([self._prefix, "sessions", _encoded(project_key)])

    def _subkeys_key(self, key: SessionListSubkeysKey) -> str:
        return ":".join(
            [
                self._prefix,
                "subkeys",
                _encoded(key["project_key"]),
                _encoded(key["session_id"]),
            ]
        )


async def run_query(
    *,
    workspace: Path,
    prompt: str,
    store: SessionStore,
    resume: str | None = None,
) -> ResultMessage:
    options = QoderAgentOptions(
        auth=access_token_from_env(),
        cwd=workspace,
        tools=[],
        max_turns=1,
        model="auto",
        session_store=store,
        resume=resume,
    )
    result: ResultMessage | None = None
    async for message in query(prompt=prompt, options=options):
        if isinstance(message, SDKMirrorErrorMessage):
            raise RuntimeError(
                f"SessionStore could not persist {message.key}: {message.error}"
            )
        if isinstance(message, ResultMessage):
            if message.subtype != "success":
                raise RuntimeError("\n".join(message.errors or [message.subtype]))
            result = message

    if result is None:
        raise RuntimeError("The query ended without a success result.")
    return result


async def run(workspace: Path) -> None:
    client = Redis.from_url(
        os.getenv("QODER_SAMPLE_REDIS_URL", "redis://127.0.0.1:6379"),
        decode_responses=True,
    )
    store = RedisSessionStore(client)
    marker = f"session-storage-{uuid.uuid4()}"
    session_id: str | None = None

    try:
        print("[host-a] Starting a session with Redis storage.")
        first = await run_query(
            workspace=workspace,
            store=store,
            prompt=(
                f"Remember this exact deployment marker: {marker}. "
                "Reply only that it is stored."
            ),
        )
        session_id = first.session_id
        delete_session(session_id, str(workspace))
        print(f"[host-a] Stored session {session_id}; local transcript deleted.")

        print("[host-b] Starting without host A's local transcript.")
        resumed = await run_query(
            workspace=workspace,
            store=store,
            resume=session_id,
            prompt=(
                "What exact deployment marker did I ask you to remember? "
                "Reply only with the marker."
            ),
        )
        print(f"[host-b] Resumed session {resumed.session_id}.")
        print(f"[host-b] Agent response: {resumed.result}")

        if marker not in (resumed.result or ""):
            raise RuntimeError(
                "The resumed response did not contain the stored marker."
            )
        print("[app] External session handoff verified.")
    finally:
        try:
            if session_id:
                if get_session_info(session_id, str(workspace)) is not None:
                    delete_session(session_id, str(workspace))
                await delete_session_via_store(store, session_id, workspace)
                print("[app] Deleted sample transcript.")
        finally:
            await client.aclose()


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("workspace", nargs="?", type=Path, default=Path.cwd())
    workspace = cast(Path, parser.parse_args().workspace).resolve()
    asyncio.run(run(workspace))


if __name__ == "__main__":
    try:
        main()
    except (RuntimeError, OSError, ValueError) as error:
        raise SystemExit(str(error)) from error

File: typescript/external-session-storage/redis-sample.ts
import { Buffer } from "node:buffer";
import { randomUUID } from "node:crypto";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

import {
  accessTokenFromEnv,
  deleteSession,
  getSessionInfo,
  query,
  type SDKResultSuccess,
  type SessionKey,
  type SessionStore,
  type SessionStoreEntry,
} from "@qoder-ai/qoder-agent-sdk";
import { Redis } from "ioredis";

const MAIN_TRANSCRIPT = "main";

function encoded(value: string): string {
  return Buffer.from(value, "utf8").toString("base64url");
}

async function execute(
  transaction: ReturnType<Redis["multi"]>,
): Promise<void> {
  const results = await transaction.exec();
  if (results === null) {
    throw new Error("The Redis transaction was discarded.");
  }
  for (const [error] of results) {
    if (error) throw error;
  }
}

/**
 * Redis reference adapter.
 *
 * Lists hold transcript entries in append order. A sorted set indexes main
 * sessions by modification time, and a set indexes child transcript paths.
 */
export class RedisSessionStore implements SessionStore {
  private readonly prefix: string;

  constructor(
    private readonly client: Redis,
    prefix = "qoder:samples:session-storage",
  ) {
    this.prefix = prefix.replace(/:+$/, "");
  }

  async append(
    key: SessionKey,
    entries: SessionStoreEntry[],
  ): Promise<void> {
    if (entries.length === 0) return;

    const transaction = this.client.multi();
    transaction.rpush(
      this.entriesKey(key),
      ...entries.map((entry) => JSON.stringify(entry)),
    );
    if (key.subpath === undefined) {
      transaction.zadd(
        this.sessionsKey(key.projectKey),
        Date.now(),
        key.sessionId,
      );
    } else {
      transaction.sadd(this.subkeysKey(key), key.subpath);
    }
    await execute(transaction);
  }

  async load(key: SessionKey): Promise<SessionStoreEntry[] | null> {
    const values = await this.client.lrange(this.entriesKey(key), 0, -1);
    if (values.length === 0) return null;
    return values.map((value) => JSON.parse(value) as SessionStoreEntry);
  }

  async listSessions(
    projectKey: string,
  ): Promise<Array<{ sessionId: string; mtime: number }>> {
    const values = await this.client.zrevrange(
      this.sessionsKey(projectKey),
      0,
      -1,
      "WITHSCORES",
    );
    const sessions: Array<{ sessionId: string; mtime: number }> = [];
    for (let index = 0; index < values.length; index += 2) {
      sessions.push({
        sessionId: values[index]!,
        mtime: Number(values[index + 1]),
      });
    }
    return sessions;
  }

  async listSubkeys(
    key: Omit<SessionKey, "subpath">,
  ): Promise<string[]> {
    return (await this.client.smembers(this.subkeysKey(key))).sort();
  }

  async delete(key: SessionKey): Promise<void> {
    if (key.subpath !== undefined) {
      const transaction = this.client
        .multi()
        .del(this.entriesKey(key))
        .srem(this.subkeysKey(key), key.subpath);
      await execute(transaction);
      return;
    }

    const subpaths = await this.client.smembers(this.subkeysKey(key));
    const childKeys = subpaths.map((subpath) =>
      this.entriesKey({ ...key, subpath }),
    );
    const transaction = this.client
      .multi()
      .del(this.entriesKey(key), this.subkeysKey(key), ...childKeys)
      .zrem(this.sessionsKey(key.projectKey), key.sessionId);
    await execute(transaction);
  }

  private entriesKey(key: SessionKey): string {
    const transcript = key.subpath !== undefined
      ? `child:${encoded(key.subpath)}`
      : MAIN_TRANSCRIPT;
    return [
      this.prefix,
      "entries",
      encoded(key.projectKey),
      encoded(key.sessionId),
      transcript,
    ].join(":");
  }

  private sessionsKey(projectKey: string): string {
    return [this.prefix, "sessions", encoded(projectKey)].join(":");
  }

  private subkeysKey(key: Omit<SessionKey, "subpath">): string {
    return [
      this.prefix,
      "subkeys",
      encoded(key.projectKey),
      encoded(key.sessionId),
    ].join(":");
  }
}

async function runQuery(options: {
  workspace: string;
  prompt: string;
  store: SessionStore;
  resume?: string;
}): Promise<SDKResultSuccess> {
  const stream = query({
    prompt: options.prompt,
    options: {
      auth: accessTokenFromEnv(),
      cwd: options.workspace,
      tools: [],
      maxTurns: 1,
      model: "auto",
      sessionStore: options.store,
      ...(options.resume ? { resume: options.resume } : {}),
    },
  });

  let result: SDKResultSuccess | undefined;
  try {
    for await (const message of stream) {
      if (message.type === "system" && message.subtype === "mirror_error") {
        throw new Error(
          `SessionStore could not persist ${JSON.stringify(message.key)}: ${message.error}`,
        );
      }
      if (message.type === "result") {
        if (message.subtype !== "success") {
          throw new Error(message.errors?.join("\n") || message.subtype);
        }
        result = message;
      }
    }
  } finally {
    await stream.close();
  }

  if (!result) throw new Error("The query ended without a success result.");
  return result;
}

export async function run(workspace: string): Promise<void> {
  const client = new Redis(
    process.env["QODER_SAMPLE_REDIS_URL"] ?? "redis://127.0.0.1:6379",
  );
  const store = new RedisSessionStore(client);
  const marker = `session-storage-${randomUUID()}`;
  let sessionId: string | undefined;

  try {
    console.log("[host-a] Starting a session with Redis storage.");
    const first = await runQuery({
      workspace,
      store,
      prompt: `Remember this exact deployment marker: ${marker}. Reply only that it is stored.`,
    });
    sessionId = first.session_id;
    await deleteSession(sessionId, { dir: workspace });
    console.log(
      `[host-a] Stored session ${sessionId}; local transcript deleted.`,
    );

    console.log("[host-b] Starting without host A's local transcript.");
    const resumed = await runQuery({
      workspace,
      store,
      resume: sessionId,
      prompt:
        "What exact deployment marker did I ask you to remember? Reply only with the marker.",
    });
    console.log(`[host-b] Resumed session ${resumed.session_id}.`);
    console.log(`[host-b] Agent response: ${resumed.result}`);

    if (!resumed.result.includes(marker)) {
      throw new Error("The resumed response did not contain the stored marker.");
    }
    console.log("[app] External session handoff verified.");
  } finally {
    try {
      if (sessionId) {
        const localSession = await getSessionInfo(sessionId, {
          dir: workspace,
        });
        if (localSession) {
          await deleteSession(sessionId, { dir: workspace });
        }
        await deleteSession(sessionId, {
          dir: workspace,
          sessionStore: store,
        });
        console.log("[app] Deleted sample transcript.");
      }
    } finally {
      await client.quit();
    }
  }
}

async function main(): Promise<void> {
  await run(resolve(process.argv[2] ?? process.cwd()));
}

if (
  process.argv[1] &&
  import.meta.url === pathToFileURL(process.argv[1]).href
) {
  main().catch((error: unknown) => {
    console.error(error instanceof Error ? error.message : error);
    process.exitCode = 1;
  });
}
